File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:


        def dfs_check(node,low,high):

            if not node:

                return True

            if not ( low < node.val < high):

                return False

            
            return dfs_check(node.left,low,node.val) and dfs_check(node.right,node.val,high)


        return dfs_check(root,float('-inf'), float('inf'))


            

            

            

            
        

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func buildTree(preorder []int, inorder []int) *TreeNode {

	inorderMap := make(map[int]int)
	currRootInd := 0

	for i, val := range inorder {

		inorderMap[val] = i
	}

	var build func(left, right int) *TreeNode 

	build = func(left, right int) *TreeNode{

		if left > right {
			return nil
		}

		root := &TreeNode{}
		root.Val = preorder[currRootInd]

		currRootInd++

		mid := inorderMap[root.Val]

		root.Left = build(left, mid-1)
		root.Right = build(mid+1,right)

		return root
	}
    

	return build(0,len(inorder)-1)
}
